File: chroma-league/src/main/java/net/booone/chromaleague/tasks/FetchNewEventsTask.java
package net.booone.chromaleague.tasks;

import net.booone.chromaleague.rest.eventdata.Event;
import net.booone.chromaleague.rest.http.LeagueHttpClients;
import net.booone.chromaleague.rest.http.handlers.EventsResponseHandler;
import net.booone.chromaleague.state.RunningState;

import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class FetchNewEventsTask implements Runnable {
    public static final String URL = "https://127.0.0.1:2999/liveclientdata/eventdata";
    private static final Logger LOGGER = Logger.getLogger(FetchNewEventsTask.class.getName());
    private final EventsResponseHandler eventsResponseHandler = new EventsResponseHandler();

    @Override
    public void run() {
        try {
            LeagueHttpClients.getNonBlockingResponse(URL, eventsResponseHandler)
                    .ifPresent(FetchNewEventsTask::collectUnprocessedEvents);
        } catch (Exception ex) {
            LOGGER.log(Level.SEVERE, "ERROR", ex);
        }
    }

    static void collectUnprocessedEvents(List<Event> events) {
        if (!RunningState.getRunningGame().getValue()) {
            waitForGameStart(events);
        } else if (RunningState.getGameState().getActivePlayer() != null && RunningState.getGameState().getPlayerList() != null) {
            final List<Event> unprocessedEvents = RunningState.getGameState().getEventData().getUnprocessedEvents(events);
            if (!unprocessedEvents.isEmpty()) {
                if (hasPlayerReconnected(unprocessedEvents)) {
                    final double gameTimeForReconnection = RunningState.getGameState().getGameStats().gameTime();
                    new EventDataProcessor(gameTimeForReconnection).processNewEvents(unprocessedEvents);
                } else {
                    new EventDataProcessor().processNewEvents(unprocessedEvents);
                    EventAnimationProcessor.processNewEvents(unprocessedEvents);
                }
                RunningState.getGameState().getEventData().addProcessedEvents(unprocessedEvents);
            }
        }
    }

    private static boolean hasPlayerReconnected(List<Event> events) {
        return events.get(events.size() - 1).EventTime() - events.get(0).EventTime() > 10;
    }

    private static void waitForGameStart(List<Event> events) {
        if (!events.isEmpty()) {
            new EventDataProcessor().processNewEvents(events.subList(0, 1));
        }
    }
}

File: src/tasks/FetchNewEventsTask.ts
import {getNonBlockingResponse} from '../rest/http/LeagueHttpClients.js';
import type {Event} from '../rest/eventdata/Event.js';
import type {Events} from '../rest/eventdata/Events.js';
import {RunningState} from '../state/RunningState.js';
import {EventDataProcessor} from './EventDataProcessor.js';
import {processNewEventAnimations} from './EventAnimationProcessor.js';

const URL = 'https://127.0.0.1:2999/liveclientdata/eventdata';

export async function fetchNewEventsTask(): Promise<void> {
    try {
        const response = await getNonBlockingResponse<Events>(URL);
        if (response) {
            collectUnprocessedEvents(response.Events);
        }
    } catch (ex) {
        console.error('ERROR', ex);
    }
}

export function collectUnprocessedEvents(events: Event[]): void {
    if (!RunningState.getRunningGame().getValue()) {
        waitForGameStart(events);
    } else {
        const gameState = RunningState.getGameState();
        if (gameState.activePlayer !== null && gameState.playerList !== null) {
            const unprocessedEvents = gameState.eventData.getUnprocessedEvents(events);
            if (unprocessedEvents.length > 0) {
                if (hasPlayerReconnected(unprocessedEvents, gameState.eventData.getLastProcessedEventTime())) {
                    const gameTimeForReconnection = gameState.gameStats?.gameTime ?? 0;
                    new EventDataProcessor(gameTimeForReconnection).processNewEvents(unprocessedEvents);
                } else {
                    new EventDataProcessor().processNewEvents(unprocessedEvents);
                    processNewEventAnimations(unprocessedEvents);
                }
                gameState.eventData.addProcessedEvents(unprocessedEvents);
            }
        }
    }
}

/** Events older than 10s on first reconnect are considered already processed */
const INITIAL_RECONNECT_GAP_SECONDS = 10;
/** Events older than 5 minutes on a later reconnect within the same session are skipped */
const SESSION_RECONNECT_GAP_SECONDS = 300;

function hasPlayerReconnected(events: Event[], lastProcessedEventTime: number | null): boolean {
    if (lastProcessedEventTime === null) {
        return events[events.length - 1].EventTime - events[0].EventTime > INITIAL_RECONNECT_GAP_SECONDS;
    }
    return events[0].EventTime - lastProcessedEventTime > SESSION_RECONNECT_GAP_SECONDS;
}

function waitForGameStart(events: Event[]): void {
    if (events.length > 0) {
        new EventDataProcessor().processNewEvents(events.slice(0, 1));
    }
}
